ESP32-S3: Failed to resolve component 'esp_adc_cal' with ESP-IDF 5.4.x (breaks voice assistants)

### The problem

ESP32-S3 voice assistant configurations fail to compile starting with ESPHome 2025.6.0 due to a missing esp_adc_cal dependency.

The error occurs during CMake configuration when ESP-IDF 5.4.x is auto-selected. The esp_adc_cal component was deprecated/removed for ESP32-S3 in ESP-IDF 5.x, but something in the component dependency chain (esp_peripherals) still requires it.

This affects multiple ESP32-S3 voice assistant hardware including Sonocotta Esparagus Echo Duo and likely others.

Working: ESPHome 2025.5.2 and earlier (auto-selects ESP-IDF 5.2.x)
Broken: ESPHome 2025.6.0 through 2025.9.3 (auto-selects ESP-IDF 5.4.2)

### Which version of ESPHome has the issue?

2025.9.3 (also affects 2025.6.0 through 2025.9.x)

### What type of installation are you using?

Docker

### What platform are you using?

ESP32

### Component causing the issue

esp-idf / framework

### YAML Config

```yaml
substitutions:
  name: esparagus-duo
  friendly_name: Esparagus Duo

esphome:
  name: ${name}
  friendly_name: ${friendly_name}
  platformio_options:
    build_flags: -DBOARD_HAS_PSRAM
    board_build.flash_mode: dio

esp32:
  board: esp32-s3-devkitc-1
  variant: ESP32S3
  flash_size: 16MB
  framework:
    type: esp-idf
    # No version specified - auto-selects ESP-IDF 5.4.2 in 2025.9.3
    sdkconfig_options:
      CONFIG_ESP32_S3_BOX_BOARD: "y"

logger:
api:
ota:
  platform: esphome

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

psram:
  mode: octal
  speed: 80MHz

i2s_audio:
  - id: i2s_in
    i2s_lrclk_pin: GPIO7 
    i2s_bclk_pin: GPIO16
  - id: i2s_out
    i2s_lrclk_pin: GPIO8
    i2s_bclk_pin: GPIO18

microphone:
  - platform: i2s_audio
    id: voice_mic
    i2s_audio_id: i2s_in
    i2s_din_pin: GPIO15
    adc_type: external
    pdm: false
    channel: right
    sample_rate: 16000
    bits_per_sample: 32bit

speaker:
  - platform: i2s_audio
    id: voice_speaker
    i2s_audio_id: i2s_out
    i2s_dout_pin: GPIO17
    dac_type: external
    sample_rate: 48000
    bits_per_sample: 16bit

voice_assistant:
  microphone: voice_mic
  speaker: voice_speaker
```

### Anything in the logs that might be useful for us?

```txt
Reading CMake configuration...
-- Building ESP-IDF components for target esp32s3
Processing 3 dependencies:
  [1/3] espressif/mdns (1.8.2)
  [2/3] zorxx/multipart-parser (1.0.1)
  [3/3] idf (5.4.2)

CMake Error at /cache/platformio/packages/framework-espidf@src-9531ca4433dfe754a3e12887170ce9d5/tools/cmake/build.cmake:328 (message):
  Failed to resolve component 'esp_adc_cal' required by component
  'esp_peripherals': unknown name.
Call Stack (most recent call first):
  /cache/platformio/packages/framework-espidf@src-9531ca4433dfe754a3e12887170ce9d5/tools/cmake/build.cmake:371 (__build_resolve_and_add_req)
  /cache/platformio/packages/framework-espidf@src-9531ca4433dfe754a3e12887170ce9d5/tools/cmake/build.cmake:378 (__build_expand_requirements)
  /cache/platformio/packages/framework-espidf@src-9531ca4433dfe754a3e12887170ce9d5/tools/cmake/build.cmake:675 (__build_expand_requirements)
  /cache/platformio/packages/framework-espidf@src-9531ca4433dfe754a3e12887170ce9d5/tools/cmake/project.cmake:718 (idf_build_process)
  CMakeLists.txt:3 (project)

========================= [FAILED] Took 161.16 seconds =========================
```

### Additional information

Affected Hardware:
- Sonocotta Esparagus Echo Duo (confirmed)
- Likely other ESP32-S3 voice assistant devices

Workaround:
Downgrade to ESPHome 2025.5.2, which auto-selects ESP-IDF 5.2.x instead of 5.4.x

The esp_adc_cal component was deprecated for ESP32-S3 in ESP-IDF 5.x and fully 
removed in 5.4.x. The component dependency chain needs updating to remove the 
esp_adc_cal requirement for ESP32-S3 targets.

Relevant ESP-IDF documentation:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/migration-guides/release-5.x/5.0/peripherals.html#adc-calibration

Hardware manufacturer confirmed last working version was ESPHome 2025.5.1:
https://github.com/sonocotta/esparagus-echo/issues/[issue_number]

just wanted to add this seems to affect more and more general cases as well. 
I started playing with a ESP32-C6 a few weeks ago. nothing fancy, some i2c's, a com ... worked fine until the upgrade to 2025.9.3 (from 2025.9.1)

now even if I try to update/install a completely clean config file, with no sensors at all I get

```
Reading CMake configuration...
-- git rev-parse returned 'fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).'
-- Building ESP-IDF components for target esp32c6
-- Configuring incomplete, errors occurred!

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/framework-espidf/tools/cmake/build.cmake:620 (message):
  Traceback (most recent call last):

    File "<frozen runpy>", line 189, in _run_module_as_main
    File "<frozen runpy>", line 112, in _get_module_details
    File "/root/.platformio/penv/.espidf-5.4.2/lib/python3.12/site-packages/idf_component_manager/__init__.py", line 6, in <module>
      from idf_component_tools.environment import ComponentManagerSettings
    File "/root/.platformio/penv/.espidf-5.4.2/lib/python3.12/site-packages/idf_component_tools/__init__.py", line 9, in <module>
      if ComponentManagerSettings().DEBUG_MODE:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/root/.platformio/penv/.espidf-5.4.2/lib/python3.12/site-packages/pydantic_settings/main.py", line 193, in __init__
      super().__init__(
    File "/root/.platformio/penv/.espidf-5.4.2/lib/python3.12/site-packages/pydantic/main.py", line 250, in __init__
      validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/root/.platformio/penv/.espidf-5.4.2/lib/python3.12/site-packages/idf_component_tools/environment.py", line 148, in fallback_to_default
      if field.validation_alias:
         ^^^^^^^^^^^^^^^^^^^^^^

  AttributeError: 'NoneType' object has no attribute 'validation_alias'

Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf/tools/cmake/project.cmake:718 (idf_build_process)
  CMakeLists.txt:3 (project)```

--- a/esphome/components/esp32/__init__.py
+++ b/esphome/components/esp32/__init__.py
@@ -877,6 +877,11 @@ async def to_code(config):
     for clean_var in ("IDF_PATH", "IDF_TOOLS_PATH"):
         os.environ.pop(clean_var, None)
 
+    # Set the location of the IDF component cache
+    os.environ["IDF_COMPONENT_CACHE_PATH"] = str(
+        CORE.relative_internal_path(".espressif")
+    )
+
     add_extra_script(
         "post",
         "post_build.py",

--- a/esphome/components/esp32/__init__.py
+++ b/esphome/components/esp32/__init__.py
@@ -877,7 +877,7 @@ async def to_code(config):
     for clean_var in ("IDF_PATH", "IDF_TOOLS_PATH"):
         os.environ.pop(clean_var, None)
 
-    # Set the location of the IDF component cache
+    # Set the location of the IDF component manager cache
     os.environ["IDF_COMPONENT_CACHE_PATH"] = str(
         CORE.relative_internal_path(".espressif")
     )

